File: solutions/python/Trapping Rain Water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n=len(height)
        left=[0]*n
        right=[0]*n
        water=0
        left[0]=height[0]
        for i in range(1,n):
            left[i]=max(height[i],left[i-1])

        right[n-1]=height[n-1]
        for i in range(n-2,-1,-1):
            right[i]=max(height[i],right[i+1])

        for i in range(0,n):
            water=water+min(left[i],right[i])-height[i]
        return water

File: solutions/cpp/Trapping Rain Water.cpp
#include <vector>
using namespace std;
class Solution
{
public:
    int min(int a, int b)
    {
        return a < b ? a : b;
    }
    int max(int a, int b)
    {
        return a > b ? a : b;
    }
    int trap(vector<int> &height)
    {
        vector<int> left, right;
        int water = 0;
        for (int i = 0; i < height.size(); i++)
        {
            left.push_back(0);
            right.push_back(0);
        }
        left[0] = height[0];
        for (int i = 1; i < height.size(); i++)
            left[i] = max(height[i], left[i - 1]);

        right[height.size() - 1] = height[height.size() - 1];
        for (int j = height.size() - 2; j >= 0; j--)
            right[j] = max(right[j + 1], height[j]);

        for (int i = 0; i < height.size(); i++)
            water = water + min(left[i], right[i]) - height[i];
        return water;
    }
};